src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp

```diff
@@ -1652,9 +1652,56 @@ static void continuation_enter_cleanup(MacroAssembler* masm) {
 #endif
 
   __ ld_ptr(tmp1, ContinuationEntry::parent_cont_fastpath_offset(), R1_SP);
+  __ st_ptr(tmp1, JavaThread::cont_fastpath_offset(), R16_thread);
+
+  if (CheckJNICalls) {
+    // Check if this is a virtual thread continuation
+    Label L_skip_vthread_code;
+    __ lwz(R0, in_bytes(ContinuationEntry::flags_offset()), R1_SP);
+    __ cmpwi(CCR0, R0, 0);
+    __ beq(CCR0, L_skip_vthread_code);
+
+    // If the held monitor count is > 0 and this vthread is terminating then
+    // it failed to release a JNI monitor. So we issue the same log message
+    // that JavaThread::exit does.
+    __ ld(R0, in_bytes(JavaThread::jni_monitor_count_offset()), R16_thread);
+    __ cmpdi(CCR0, R0, 0);
+    __ beq(CCR0, L_skip_vthread_code);
+
+    // Save return value potentially containing the exception oop
+    Register ex_oop = R15_esp;   // nonvolatile register
+    __ mr(ex_oop, R3_RET);
+    __ call_VM_leaf(CAST_FROM_FN_PTR(address, SharedRuntime::log_jni_monitor_still_held));
+    // Restore potentional return value
+    __ mr(R3_RET, ex_oop);
+
+    // For vthreads we have to explicitly zero the JNI monitor count of the carrier
+    // on termination. The held count is implicitly zeroed below when we restore from
+    // the parent held count (which has to be zero).
```

Comment:
In the general case the parent held count might be non-zero, but with vthreads and pinning it has to be zero. What we need to do with this code is ensure the held-count and jni-count are in sync: held-count >= jni-count. If the held-count could be non-zero here then we would not know what value to set in the jni-count to make sure the relationship is correct. It is only because we know the held-count has to be zero that we know we have to set the jni-count to zero.

Reply:
> In the general case the parent held count might be non-zero, but with vthreads and pinning it has to be zero.

With Patricios work in the loom repo we can reach here also when the vthread owns a Java monitor. I see that if the monitor was entered using JNI that this still prevents context switching (https://github.com/openjdk/loom/blob/09e4329fea5e3908855fa0881f156b7fa300a533/src/hotspot/share/runtime/continuationFreezeThaw.cpp#L1839-L1840). I somehow didn't expect this but after thinking about it twice I think it makes sense.

So reaching here with a non-zero JNI monitor count this means that the vthread is terminating. The following line could be improved to reflect this better if you want.

```
If the held monitor count is > 0 and this vthread is terminating then
```